Refactor simulation interface to default FIP-81 behavior and remove explicit gamma handling

luca-nik opened the pull request:
##  Refactor simulation interface to default FIP-81 behavior and remove explicit gamma handling

###  Summary

This PR refactors the `sim.run_sim` function to **default to FIP-81 behavior**.

It removes the need to explicitly provide `gamma` and `gamma_weight_type`, while still maintaining compatibility with FIP-81 pledge mechanics.

---

###  Changes Introduced

- **FIP-81 behavior is now defaulted inside `sim.run_sim`:**
  - `gamma` is internally constructed using a smooth ramp from 1.0 → 0.7 after FIP-81 activation.
  - `gamma_weight_type` is fixed to `0`, i.e., **arithmetic weighting**.

- **User-facing API is simplified:**
  ####  Before:

```@python
  simulation_results = sim.run_sim(
      rbp,
      rr,
      fpr,
      lock_target,
      start_date,
      current_date,
      forecast_length_days,
      sector_duration_days,
      offline_data,
      gamma=gamma_trajectory,
      gamma_weight_type=0,
      use_available_supply=False,
  )
```


####  Now:

```@python
simulation_results = sim.run_sim(
    rbp,
    rr,
    fpr,
    lock_target,
    start_date,
    current_date,
    forecast_length_days,
    sector_duration_days,
    offline_data,
    use_available_supply=False,
)
```

* **Updated logic in `sim.py`:**

  * A default gamma trajectory is computed using:

    ```python
    full_gamma_vec = create_gamma_trajectory(current_date, forecast_length, len(historical_renewal_rate))
    ```

    and passed internally to the supply forecast model.

* **`locking.weighted_consensus(...)` updated to fixed arithmetic form:**

  ```python
  @jax.jit
  def weighted_consensus(w, simple_consensus_pledge, baseline_consensus_pledge):
      return (1 - w) * simple_consensus_pledge + w * baseline_consensus_pledge
  ```

---

###  Validation

[A debugging folder](https://drive.google.com/drive/folders/1etBZjy_PruV6lExrO0WXkoEH4FRcG0Zy) contains test scripts.

These scripts compare main vs fip81 branches and validate output deltas (`network_locked`) across simulation runs.

To run these scripts:

1. Download the folder from the link above
2. In `run_and_compare.sh`, replace the path `/Users/luca/programmi/cel/mechafil/mechafil-jax` with the path to the local clone of the `mechafil-jax` repo
3. Run:
```bash
./run_and_compare.sh
```

The script will:

1. Automatically check-out the correct branches
2. Run both versions of the simulation
3. Compare the results via a plot

kkarrancsu commented on mechafil_jax/locking.py:
```diff
@@ -186,37 +174,9 @@ def compute_new_pledge_for_added_power(
 
 @jax.jit
-def arithmetic_fn(w, simple_consensus_pledge, baseline_consensus_pledge):
+def weighted_consensus(w, simple_consensus_pledge, baseline_consensus_pledge):
```
is it correct to understand that part of this change removes the other weighting mechanisms that were in there, but not being used?

luca-nik replied:
Yes, that's correct, it removes the other unused weighting mechanisms

kkarrancsu commented on mechafil_jax/locking.py:
```diff
@@ -264,3 +224,26 @@ def get_day_schedule_pledge_release(
+
+
+# Creates the trajectory of the gamma smoothing function for FIP-81 activation
+def create_gamma_trajectory(current_date: np.datetime64, 
```
nice - this is the same thing as was in the notebooks IIRC.  is there a way to cross-check this with what is on-chain (can be something for later), maybe we add it as an issue to the repo and then work on it when we have mroe time.

luca-nik replied:
Yes, it can be a good idea. When I have some time I'll try to see if I can set up an oracle for the current gamma value on-chain.

kkarrancsu reviewed (changes requested):
looks good, my main feedback is that we don't want a standalone separate folder with code that we can run to test.

it should be baked into the unit tests.

also - b/c this change now defaults to fip81, the iniitial unit tests will likely fail.  we can keep that as a separate issue and fix it with a different P/R.

luca-nik commented:
Updated the unit tests; all tests now pass.

Additionally, I addressed one of the TODOs in the unit tests, specifically the setup for data access. The original implementation was as follows:

```python
    # setup data access
    # TODO: better way to do this?
    data.setup_spacescope(os.path.join(os.environ['HOME'],'code/cel/auth/spacescope_auth.json'))
    mecha_data.setup_spacescope(os.path.join(os.environ['HOME'],'code/cel/auth/spacescope_auth.json'))
```

This has been refactored to improve flexibility and avoid hardcoding paths:
```python
# Load the .env file
load_dotenv()
auth_file_path = os.getenv('SPACESCOPE_AUTH_PATH')
        
# Setup data access
if auth_file_path:
    data.setup_spacescope(auth_file_path)
    mecha_data.setup_spacescope(auth_file_path)
    else:
        raise Exception("SPACESCOPE_AUTH_PATH environment variable is not set")
```

o facilitate this change, I introduced a private .env file, which is excluded from version control via .gitignore. An example .env file is provided as .example.env, and it contains the following structure:
```console
SPACESCOPE_AUTH_PATH=/path/to/spacescope_auth.json
```

luca-nik commented:
I have added tests to verify the new implementation of FIP81 against the previous implementation in the main branch.

To achieve this, I included the result of a calculation obtained from the original main branch as a reference. During testing, I identified that truncation errors in floating-point operations primarily caused numerical discrepancies.

After increasing the numerical precision and ensuring that all variables involved in constructing the gamma trajectory were appropriately handled with the correct precision, the tests now pass successfully.

This ensures consistency between the two implementations and addresses the floating-point precision issues that were previously causing mismatches.

The code used to generate the reference file `results_original_branch.pkl` is the following:

```python
import pickle
import numpy as np
import jax.numpy as jnp
from datetime import date, timedelta
from diskcache import Cache

import mechafil_jax.data as data
import mechafil_jax.sim as sim
import scenario_generator.utils as u
from jax import config
config.update("jax_enable_x64", True)

def run_simulation(output_filename: str):
    ###########################
    # Helper Functions
    ###########################
    def get_offline_data(start_date, current_date, end_date):
        cache = Cache("./cache_directory")
        cache_key = f"offline_data_{start_date}_{current_date}_{end_date}"
        cached_result = cache.get(cache_key)
        if cached_result is not None:
            return cached_result

        PUBLIC_AUTH_TOKEN = 'Bearer ghp_xxxxxxxx'
        offline_data = data.get_simulation_data(PUBLIC_AUTH_TOKEN, start_date, current_date, end_date)

        _, hist_rbp = u.get_historical_daily_onboarded_power(current_date - timedelta(days=180), current_date)
        _, hist_rr = u.get_historical_renewal_rate(current_date - timedelta(days=180), current_date)
        _, hist_fpr = u.get_historical_filplus_rate(current_date - timedelta(days=180), current_date)
        smoothed_last_historical_rbp = float(np.median(hist_rbp[-30:]))
        smoothed_last_historical_rr = float(np.median(hist_rr[-30:]))
        smoothed_last_historical_fpr = float(np.median(hist_fpr[-30:]))

        result = (
            offline_data, smoothed_last_historical_rbp, smoothed_last_historical_rr,
            smoothed_last_historical_fpr, hist_rbp, hist_rr, hist_fpr
        )
        cache.set(cache_key, result)
        return result

    def create_gamma_trajectory(current_date, forecast_length_days, fip81_activation_date, ramp_len_days=365):
        gamma_target = 0.7
        days_since_activation = (current_date - fip81_activation_date).days
        gamma_slope = (1.0 - gamma_target) / ramp_len_days
        current_gamma = 1.0 - gamma_slope * days_since_activation
        print(f'current_gamma: {current_gamma}')
        remaining_days = ramp_len_days - days_since_activation
        v1 = jnp.linspace(current_gamma, gamma_target, remaining_days)
        v2 = jnp.ones(forecast_length_days - remaining_days) * gamma_target
        return jnp.concatenate([v1, v2])

    ###########################
    # Main Logic
    ###########################
    print('Loading historical data ...')
    current_date = date(2025, 7, 6)
    mo_start = max(current_date.month - 1 % 12, 1)
    start_date = date(current_date.year, mo_start, 1)
    forecast_length_days = 365 * 5
    end_date = current_date + timedelta(days=forecast_length_days)

    offline_data, smoothed_rbp, smoothed_rr, smoothed_fpr, *_ = get_offline_data(start_date, current_date, end_date)

    sector_duration_days = 540
    lock_target = 0.3
    fip81_activation_date = date(2024, 11, 21)
    gamma_trajectory = create_gamma_trajectory(current_date, forecast_length_days, fip81_activation_date)

    print(f'Current smoothed last historical RBP: {smoothed_rbp}')
    print('Running simulations ...')

    rbp = jnp.ones(forecast_length_days) * smoothed_rbp
    rr = jnp.ones(forecast_length_days) * smoothed_rr
    fpr = jnp.ones(forecast_length_days) * smoothed_fpr

    simulation_results = sim.run_sim(
        rbp, rr, fpr, lock_target, start_date, current_date,
        forecast_length_days, sector_duration_days, offline_data,
        gamma=gamma_trajectory, gamma_weight_type=0, use_available_supply=False
    )

    with open(output_filename, "wb") as f:
        pickle.dump([simulation_results], f)

    print(f"Simulation saved to {output_filename}")
```

kkarrancsu commented on test/test_fip81.py:
```diff
@@ -0,0 +1,104 @@
+
+        # Load data from original fip81 implementation
+        test_dir = os.path.dirname(__file__)  # This will get the directory of the current test file
+        file_path = os.path.join(test_dir, 'results_original_branch.pkl')
```
this is opaque, so it would be better if we can have the code which generates this pickle in the unittest file.  otherwise, it is hard to tell what the reference actually is.

kkarrancsu reviewed (changes requested):
need to remove pickle file and replace with code that generates those test vectors so that the tests are transparent.
